### PR TITLE
Handle ref returning receivers in dynamic calls

### PR DESCRIPTION
**Pull request by jjonescz**
Fixes https://github.com/dotnet/roslyn/issues/68063.

### CODE REVIEW

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LoweredDynamicOperationFactory.cs [hunk 1]
```diff
@@ -549,6 +549,9 @@ internal static RefKind GetReceiverRefKind(BoundExpression loweredReceiver)
 
             switch (loweredReceiver.Kind)
             {
+                case BoundKind.Call:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LoweredDynamicOperationFactory.cs:552]: >case BoundKind.Call: [](http://example.com/codeflow?start=16&length=20)

Are we missing handling for other interesting scenarios? For example, I do not see explicit case for fields. We probably could run into a sequence (for example if we need to reorder arguments for a call), etc.
It looks like the purpose of this method is to determine whether we can take a reference to the receiver without storing the value in a temp. It feels fragile to have a helper that is used only for the purpose of dynamic rewrite, especially taking future language changes into account. Scenarios involving dynamic rewrite are easy to miss during testing. Therefore, I think we should either switch to a different existing helper, which is used more often (```Binder.HasHome```?), or we could try always pass value type receiver by reference and let emit layer to figure out if it needs a temp for that (it looks like that shouldn't change semantics of the operation).

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LoweredDynamicOperationFactory.cs [hunk 2]
```diff
@@ -539,35 +539,20 @@ private MethodSymbol GetArgumentInfoFactory()
...
-                // base dynamic dispatch is not supported, an error has already been reported
-                case BoundKind.TypeExpression:
-                    throw ExceptionUtilities.UnexpectedValue(loweredReceiver.Kind);
-            }
-
-            return RefKind.None;
+            var hasHome = Binder.HasHome(loweredReceiver,
+                Binder.AddressKind.Writeable,
+                _factory.CurrentFunction,
+                _factory.Compilation.IsPeVerifyCompatEnabled,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LoweredDynamicOperationFactory.cs:553]: >_factory.Compilation.IsPeVerifyCompatEnabled [](http://example.com/codeflow?start=16&length=44)

I think we don't need to depend on this setting here. The most permissive value is probably be the closest match with the previous behavior.

File: src/Compilers/CSharp/Test/Semantic/Semantics/DynamicTests.cs [hunk 3]
```diff
@@ -1510,6 +1510,217 @@ public void Goo()
...
+        public void StructReceiver_Rvalue()
+        {
+            var source = """
+                class Program
+                {
+                    static void Main()
+                    {
+                        dynamic d = 1;
+                        GetS().M(d);
+                        System.Console.Write(GetS().X);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DynamicTests.cs:1662]: >System.Console.Write(GetS().X); [](http://example.com/codeflow?start=24&length=31)

It looks like this line isn't meaningful. It is guaranteed to produce the same output regardless of what the previous line is doing.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5)

**Review by jaredpar — APPROVED** [review]

**Pull request merged**